### ISSUE TITLE
Proposal GrainFactory - IdSpan-based new overloads for complex keys

### ISSUE THREAD

**Issue by cmeyertons** [dotnet/orleans#8391]
Today, you have 2 options for passing custom, complex keys into `IGrainFactory` (see #2320 and https://github.com/dotnet/orleans/issues/8377#issuecomment-1500207005)

```csharp
// option 1 - string-based
TGrainInterface GetGrain<TGrainInterface>(string primaryKey, string grainClassNamePrefix = null) where TGrainInterface : IGrainWithStringKey;

// option 2 - byte[]-based, but have to drop down to GrainId
IAddressable GetGrain(GrainId grainId);
```

There are a couple of drawbacks to these options:

1. string-based - complex types have to use a string-based serialization format (JSON, etc) which hurts performance (ids are deserialized on every grain activation, I would consider this in a hot path)
2. GrainId - this overload is not user-friendly and requires a lot of knowledge of the internals on how to create a `GrainType`, so much so that [OrleansTestKit](https://github.com/OrleansContrib/OrleansTestKit) does not even have this [method implemented](https://github.com/OrleansContrib/OrleansTestKit/blob/main/src/OrleansTestKit/TestGrainFactory.cs)

Instead of the full gambit of supporting complex keys, I propose adding an IdSpan-based method to IGrainFactory that makes working with them much easier / performant

```csharp
TGrainInterface GetGrain<TGrainInterface>(IdSpan primaryKey, string grainClassNamePrefix = null) where TGrainInterface: IGrain
IGrain GetGrain(Type grainInterfaceType, IdSpan primaryKey);
```

This would allow complex key authors to create an arbitrary `IdSpan` from a `byte[]` and be able to leverage binary-based Serialization formats for the grain key (including even the Orleans Serializer).

**Comment by ReubenBond** [comment]
Sounds good to me. My only reservation is that `IGrainFactory` seems to have too many methods already. I wonder if it could be simplified.

**Comment by ReubenBond** [comment]
@cmeyertons that might work. As long as the developer experience is no worse

**Comment by SebastianStehle** [comment]
Extension methods are hard to use with mocking and the IGrainFactory is very likely mocked a lot. Implicit conversions to IdSpan might work better.

**Comment by ReubenBond** [comment]
@SebastianStehle could you mock the underlying method which the extension calls instead, or is there something tricky about that?

**Comment by SebastianStehle** [comment]
yes, you can. But then you have to understand the implementation details. So you have a copy of the implementation in your tests. It might be acceptable for this class, but I hate it for the ILogger for example (which I have to mock in very few cases).

**Comment by cmeyertons** [comment]
Maybe we should look at OrleansTestKit for the primary example -- [TestGrainFactory](https://github.com/OrleansContrib/OrleansTestKit/blob/main/src/OrleansTestKit/TestGrainFactory.cs)

It requires implementation detail knowledge to have to cover the entire interface, knowing how the "real" GrainFactory converts Guid/longs/combos into `IdSpan` under the hood.  I also don't think a unit tester would reasonably need to mock out the extension method, because its very light in logic (really just calling IdSpan.Create w/ the correct overload).  I think in this use case, the benefit of the extension method outweighs the cost.

I'm also not sure `ILogger` is a great correlation because the extension methods have a lot of logic / abstraction and have a very different "shape" compared to the core `ILogger` interface.
